generators/dart/lists.js

```diff
@@ -38,144 +38,206 @@ Blockly.Dart['lists_create_empty'] = function(block) {
+          'tmp_x', Blockly.Variables.NAME_TYPE);
+      code += 'int ' + xVar + ' = new Math.Random().nextInt(' + list +
+          '.length);\n';
+      code += list + '.removeAt(' + xVar + ');\n';
+      return code;
+    } else {  // where == 'FROM_END'
+      if (mode == 'REMOVE') {
+        // We can use multiple statements.
+        var at = Blockly.Dart.getAdjusted(block, 'AT', 1, false,
+            Blockly.Dart.ORDER_ADDITIVE);
+        var code = cacheList();
+        code += list + '.removeAt(' + list + '.length' + ' - ' + at + ');\n';
+        return code;
 
-  if (where == 'FIRST') {
-    if (mode == 'GET') {
-      var code = list + '.first';
-      return [code, Blockly.Dart.ORDER_UNARY_POSTFIX];
-    } else if (mode == 'GET_REMOVE') {
-      var code = list + '.removeAt(0)';
-      return [code, Blockly.Dart.ORDER_UNARY_POSTFIX];
-    } else if (mode == 'REMOVE') {
-      return list + '.removeAt(0);\n';
-    }
-  } else if (where == 'LAST') {
-    if (mode == 'GET') {
-      var code = list + '.last';
-      return [code, Blockly.Dart.ORDER_UNARY_POSTFIX];
-    } else if (mode == 'GET_REMOVE') {
-      var code = list + '.removeLast()';
-      return [code, Blockly.Dart.ORDER_UNARY_POSTFIX];
-    } else if (mode == 'REMOVE') {
-      return list + '.removeLast();\n';
-    }
-  } else if (where == 'FROM_START') {
-    // Blockly uses one-based indicies.
-    if (Blockly.isNumber(at)) {
-      // If the index is a naked number, decrement it right now.
-      at = parseInt(at, 10) - 1;
-    } else {
-      // If the index is dynamic, decrement it in code.
-      at += ' - 1';
-    }
-    if (mode == 'GET') {
-      var code = list + '[' + at + ']';
-      return [code, Blockly.Dart.ORDER_UNARY_POSTFIX];
-    } else if (mode == 'GET_REMOVE') {
-      var code = list + '.removeAt(' + at + ')';
-      return [code, Blockly.Dart.ORDER_UNARY_POSTFIX];
-    } else if (mode == 'REMOVE') {
-      return list + '.removeAt(' + at + ');\n';
-    }
-  } else if (where == 'FROM_END') {
-    if (mode == 'GET') {
-      var functionName = Blockly.Dart.provideFunction_(
-          'lists_get_from_end',
-          [ 'dynamic ' + Blockly.Dart.FUNCTION_NAME_PLACEHOLDER_ +
-              '(List myList, num x) {',
-            '  x = myList.length - x;',
-            '  return myList.removeAt(x);',
-            '}']);
-      code = functionName + '(' + list + ', ' + at + ')';
-      return [code, Blockly.Dart.ORDER_UNARY_POSTFIX];
-    } else if (mode == 'GET_REMOVE' || mode == 'REMOVE') {
-      var functionName = Blockly.Dart.provideFunction_(
-          'lists_remove_from_end',
-          [ 'dynamic ' + Blockly.Dart.FUNCTION_NAME_PLACEHOLDER_ +
-              '(List myList, num x) {',
-            '  x = myList.length - x;',
-            '  return myList.removeAt(x);',
-            '}']);
-      code = functionName + '(' + list + ', ' + at + ')';
-      if (mode == 'GET_REMOVE') {
+      } else if (mode == 'GET' || mode == 'GET_REMOVE') {
```

Comment:
There is not enough common code in these two cases to share.  Split them up.
Same issue again about 90 lines down.